Take from basket

mhagel opened the pull request:
## Description

Closes: regen-network/regen-registry#816

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

- Enables Basket Take functionality from Portfolio page.
- Peripheral fix: Consolidates styles for modals that display an iframe

Key files:
web-components/src/components/form/BasketTakeForm.tsx
web-components/src/utils/locationStandard.ts
web-registry/src/pages/MyEcocredits.tsx



---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] provided a link to the relevant issue or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**.*

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles 
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

blushi commented on web-registry/src/pages/MyEcocredits.tsx:
```diff
@@ -19,24 +45,117 @@ const useStyles = makeStyles((theme: Theme) => ({
+  };
+
+  const handleTakeCredits = async (values: MsgTakeValues): Promise<void> => {
+    const msgClient = api?.msgClient;
```
close the take modal here (see submit for put)

mhagel replied:
done

blushi commented on web-components/src/components/form/BasketTakeForm.tsx:
```diff
@@ -0,0 +1,228 @@
+              availableAmount={availableTradableAmount}
+              batchDenom={basket.basketDenom}
+            />
+            <Field
+              component={CheckboxLabel}
+              type="checkbox"
+              name="retireOnTake"
+              className={styles.checkboxLabel}
+              label={
+                <Description className={styles.checkboxDescription}>
+                  Retire credits upon transfer
+                </Description>
+              }
+            />
```
if the basket `disable_auto_retire` field is false then retirement is mandatory on take so in this case we should just check the retire checkbox by default and users should not be able to change that.

See:
- docs: https://buf.build/regen/regen-ledger/docs/main:regen.ecocredit.basket.v1#regen.ecocredit.basket.v1.MsgCreate 
- the **important note** in https://github.com/regen-network/regen-registry/issues/816
- design: https://www.figma.com/file/x5vjWsddiUBzP2N13AFOPw/Toucan-like-Flow-(Baskets)?node-id=628%3A15466

mhagel replied:
Thanks! Done.

blushi commented on web-registry/src/pages/MyEcocredits.tsx:
```diff
@@ -94,6 +100,49 @@ const WrappedMyEcocredits: React.FC<WithBasketsProps> = ({ baskets }) => {
+    }
+  };
+
+  const handleTakeCredits = async (values: MsgTakeValues): Promise<void> => {
```
let's be consistent with function naming, we have `handleTakeCredits` and `basketPutSubmit`
what about `basketTakeSubmit`?

mhagel replied:
Now `basketTakeSubmit`

blushi commented on web-registry/src/pages/MyEcocredits.tsx:
```diff
@@ -72,12 +76,14 @@ const WrappedMyEcocredits: React.FC<WithBasketsProps> = ({ baskets }) => {
   >([]);
   const [basketPutOpen, setBasketPutOpen] = useState<number>(-1);
   const [isProcessingModalOpen, setIsProcessingModalOpen] = useState(false);
+  const [selectedBasketDenom, setSelectedBasketDenom] = useState('');
```
still in terms of consistency, what about just using `basketTakeOpen` (as for `basketPutOpen`)?
this works right now because we only have one type of action we can do on a basket (take) but in the future, if we want to add more CTAs and from the basket table, then we'll need to change that

mhagel replied:
Now `setBasketTakeTokens` and moved logic out of modal so it can be in web-components

haveanicedavid commented on web-components/src/components/form/BasketTakeForm.tsx:
```diff
@@ -0,0 +1,224 @@
+    onSubmit(msgTake);
+  };
+
+  const AutoSetRetirementLocation = (): JSX.Element => {
```
curious why this is a component instead of a normal effect. Is it so you have access to formik context?

mhagel replied:
Yes, I pulled this from the Formik docs. I believe the hook has to be inside the <Form> for the context to work.

haveanicedavid reviewed (approved):
few small comments, but in general LGTM. Nice work!

blushi commented on web-components/src/components/form/BasketTakeForm.tsx:
```diff
@@ -0,0 +1,224 @@
+  ): FormikErrors<CreditTakeFormValues> => {
+    let errors: FormikErrors<CreditTakeFormValues> = {};
+
+    if (!values.amount) {
+      errors.amount = requiredMessage;
+    } else if (Math.sign(values.amount) !== 1) {
+      errors.amount = invalidAmount;
+    } else if (values.amount > availableTradableAmount) {
+      errors.amount = insufficientCredits;
+    }
```
you could use `validateAmount` instead, see in `BasketPutForm`

mhagel replied:
done

mhagel replied:
added custome message for basket tokens case

mhagel commented on web-components/src/utils/locationStandard.ts:
```diff
@@ -0,0 +1,42 @@
+  countryKey?: string,
+  stateProvince?: string,
+  postalCode?: string,
+): Promise<string | undefined> => {
```
done

blushi reviewed (approved):
Pre-approving, pending small nit

blushi reviewed (changes requested):
Actually, I just realized that we should probably change the error message:
![image](https://user-images.githubusercontent.com/3149358/162900589-4019b48d-a537-4737-822d-a605962227c4.png)
 in this case, the user doesn't have enough basket tokens, not credits
cc/ @erikalogie

blushi commented:
Also the unit of this amount is wrong:
![image](https://user-images.githubusercontent.com/3149358/162900768-90d96520-c0cc-41fb-acd3-30699bfe4f58.png)

it's 10000698895 eco.uC.rNCT, ie 10000.698895 eco.C.rNCT (without the micro `u`) and **not** 10000.698895 eco.uC.rNCT

mhagel commented on web-registry/src/pages/MyEcocredits.tsx:
```diff
@@ -8,18 +8,22 @@ import ArrowDownIcon from 'web-components/lib/components/icons/ArrowDownIcon';
 import { FormValues as BasketPutFormValues } from 'web-components/lib/components/form/BasketPutForm';
+import { MsgTakeValues } from 'web-components/lib/components/form/BasketTakeForm';
 import { Option } from 'web-components/lib/components/inputs/SelectTextField';
+import { MsgTake } from '@regen-network/api/lib/generated/regen/ecocredit/basket/v1/tx';
```
oh yeah. done

mhagel commented:
>Also the unit of this amount is wrong:

>Actually, I just realized that we should probably change the error message:

Fixed and fixed.

Also added a subtitle which I didn't see before.

@blushi

mhagel commented on web-components/src/components/form/CreditRetireForm.tsx:
```diff
@@ -216,7 +230,6 @@ const CreditRetireForm: React.FC<FormProps> = ({
       {({ values, submitForm, isSubmitting, isValid, submitCount, status }) => (
         <Form>
           <CreditRetireFields
-            country={values.country}
```
This is now being pulled in with `useFormikContext`

mhagel commented on web-registry/src/hooks/useMsgClient.ts:
```diff
@@ -35,26 +45,54 @@ export default function useMsgClient(
-          const _deliverTxResponse = await api.msgClient.broadcast(txBytes);
-          setDeliverTxResponse(_deliverTxResponse);
-          handleTxDelivered();
+          if (onBroadcast) onBroadcast();
```
@blushi added this, to give us optional modal-closing after Keplr signing is confirmed.

blushi commented on web-registry/src/hooks/useMsgClient.ts:
```diff
@@ -4,8 +4,17 @@ import { StdFee, DeliverTxResponse } from '@cosmjs/stargate';
 import { useLedger } from '../ledger';
 import { Sender } from '../lib/wallet';
 
+interface MessageData {
```
I think that'd be a better naming since fee and memo aren't really related to messages
```suggestion
interface TxData {
```

mhagel replied:
done

blushi commented on web-registry/src/hooks/useMsgClient.ts:
```diff
@@ -4,8 +4,17 @@ import { StdFee, DeliverTxResponse } from '@cosmjs/stargate';
 type MsgClientType = {
-  signAndBroadcast: (msgs: any, fee?: StdFee, memo?: string) => Promise<void>;
+  signAndBroadcast: (
+    message: MessageData,
```
ditto
```suggestion
    txData: TxData,
```

blushi reviewed (approved):
let's keep the original casing:
![image](https://user-images.githubusercontent.com/3149358/163131519-7f8e2069-e7a4-40aa-a7aa-01ae85112ca5.png)

pre-approving, pending small changes

mhagel commented on web-components/src/components/form/BasketTakeForm.tsx:
```diff
@@ -0,0 +1,223 @@
+    const errAmount = validateAmount(
+      availableTradableAmount,
+      values.amount,
+      `You don't have enough basket tokens`,
```
eh, I'll leave this as is for now, in case we think of some other messages.

mhagel commented:
@Mergifyio backport redwood hambach
